addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,434 @@
+	pkgi, ok := o.(*kapppkgiv1alpha1.PackageInstall)
+	if !ok {
+		r.Log.Error(errors.New("invalid type"),
+			"Expected to receive PackageInstall resource",
+			"actualType", fmt.Sprintf("%T", o))
+		return nil
+	}
+
+	log := r.Log.WithValues("pkgi-name", pkgi.Name)
+	log.V(4).Info("Mapping PackageInstalls to cluster")
+
+	clusterNamespace, clusterName := r.getClusterNamespacedName(pkgi)
+	if clusterNamespace == "" || clusterName == "" {
+		return nil
+	}
+
+	// TODO: check if the following check is needed here considering that we check it in PkgiStatusChanged predicate
+	isManaged, err := r.isPackageManaged(clusterName, pkgi.Name, clusterNamespace)
+	if err != nil || !isManaged {
+		return nil
+	}
+
+	return []ctrl.Request{{NamespacedName: client.ObjectKey{Namespace: clusterNamespace, Name: clusterName}}}
+}
+
+// PkgiStatusChanged returns a predicate.Predicate that filters pkgi objects which their status has changed
+func (r *PackageInstallStatusReconciler) PkgiStatusChanged(log logr.Logger) predicate.Funcs {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return r.processPkgiStatus(e.Object, log.WithValues("predicate", "createEvent"))
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return r.processPkgiStatus(e.ObjectNew, log.WithValues("predicate", "updateEvent"))
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return r.processPkgiStatus(e.Object, log.WithValues("predicate", "deleteEvent"))
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return r.processPkgiStatus(e.Object, log.WithValues("predicate", "genericEvent"))
+		},
+	}
+}
+
+// processPkgiStatus returns true if pkgi status should be processed.
+// pkgi status can be processed if it is a managed package and exists in the list of packages defined in ClusterBootstrap
+func (r *PackageInstallStatusReconciler) processPkgiStatus(o client.Object, log logr.Logger) bool {
+	var pkgi *kapppkgiv1alpha1.PackageInstall
+	switch obj := o.(type) {
+	case *kapppkgiv1alpha1.PackageInstall:
+		pkgi = obj
+	default:
+		//TODO: check this
+		// Defaults to true so we don't filter out other objects as the filters are global
+		log.Info("Expected object type of PackageInstall. Got object type", "actualType", fmt.Sprintf("%T", o))
+		return true
+	}
+
+	clusterNamespace, clusterName := r.getClusterNamespacedName(pkgi)
+	if clusterNamespace == "" || clusterName == "" {
+		return false
+	}
+
+	isManaged, err := r.isPackageManaged(clusterName, pkgi.Name, clusterNamespace)
+	if err != nil {
+		return false
+	}
+
+	return isManaged
+}
+
+// isPackageManaged checks if the provided PackageInstall is among the list of managed(core/additional) packages
+func (r *PackageInstallStatusReconciler) isPackageManaged(clusterName, pkgiName, pkgiNamespace string) (bool, error) {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	clusterKey := client.ObjectKey{Namespace: pkgiNamespace, Name: clusterName}
```

Comment:
suggestion: could you help add some comments help maintainer to understand why we use `pkgiNamespace/clusterName` for clusterbootstrap object ?

Reply:
And `clusterKey` is a little bit confusing, since we are fetching clusterBootstrap object.

Reply:
The use of ```pkgiNamespace``` has been a typo, in the caller I've passed ```clusterNamespace```. Will fix it, thanks for catching it!